File: companion/ble_transport.py
"""
BLE transport for Pixel Agents Bridge — connects to ESP32 via Nordic UART Service.

Uses bleak (asyncio-based BLE library) with a background event loop thread
to provide synchronous send() for the bridge's main loop.
"""

import asyncio
import threading
import time
from typing import Optional

try:
    from bleak import BleakClient, BleakScanner
except ImportError:
    print("Error: bleak is required for BLE transport. Install with: pip install bleak")
    raise

# Nordic UART Service UUIDs
NUS_SERVICE_UUID = "6e400001-b5a3-f393-e0a9-e50e24dcca9e"
NUS_RX_UUID = "6e400002-b5a3-f393-e0a9-e50e24dcca9e"  # Write to this (ESP32 receives)
NUS_TX_UUID = "6e400003-b5a3-f393-e0a9-e50e24dcca9e"  # Notify from this (ESP32 sends)

SCAN_TIMEOUT_SEC = 10.0
CONNECT_TIMEOUT_SEC = 10.0

# Manufacturer data company ID (must match firmware BLE_MFG_COMPANY_ID)
MFG_COMPANY_ID = 0xFFFF


class BleTransport:
    """BLE transport using Nordic UART Service.

    Runs bleak's asyncio event loop on a background thread.
    Exposes synchronous connect/send/disconnect for the bridge.
    """

    def __init__(self, device_name: str = "PixelAgents"):
        self._device_name = device_name
        self._client: Optional[BleakClient] = None
        self._connected = False
        self._loop: Optional[asyncio.AbstractEventLoop] = None
        self._thread: Optional[threading.Thread] = None
        self._start_event_loop()

    def _start_event_loop(self):
        """Start the asyncio event loop on a background thread."""
        self._loop = asyncio.new_event_loop()
        self._thread = threading.Thread(target=self._run_loop, daemon=True)
        self._thread.start()

    def _run_loop(self):
        """Run the asyncio event loop (background thread)."""
        asyncio.set_event_loop(self._loop)
        self._loop.run_forever()

    def _run_coro(self, coro, timeout: float = 30.0):
        """Run an async coroutine from the synchronous context."""
        future = asyncio.run_coroutine_threadsafe(coro, self._loop)
        return future.result(timeout=timeout)

    @property
    def is_connected(self) -> bool:
        return self._connected

    def scan(self) -> Optional[str]:
        """Scan for the PixelAgents device. Returns BLE address or None."""
        devices = self.scan_devices()
        if devices:
            addr, name, pin = devices[0]
            return addr
        return None

    def scan_devices(self) -> list:
        """Scan for NUS devices. Returns list of (address, name, pin) tuples."""
        return self._run_coro(self._scan_devices_async(), timeout=SCAN_TIMEOUT_SEC + 5)

    async def _scan_devices_async(self) -> list:
        print("Scanning for BLE devices (NUS service)...")
        devices = await BleakScanner.discover(
            timeout=SCAN_TIMEOUT_SEC,
            service_uuids=[NUS_SERVICE_UUID],
            return_adv=True
        )
        results = []
        for addr, (device, adv_data) in devices.items():
            pin = self._extract_pin(adv_data)
            name = adv_data.local_name or device.name or "Unknown"
            pin_str = f", PIN={pin}" if pin else ""
            print(f"  Found: {name} at {addr}{pin_str}")
            results.append((addr, name, pin))
        if not results:
            print("No NUS devices found.")
        return results

    @staticmethod
    def _extract_pin(adv_data) -> Optional[int]:
        """Extract PIN from manufacturer data in advertisement."""
        mfg = adv_data.manufacturer_data
        if not mfg:
            return None
        payload = mfg.get(MFG_COMPANY_ID)
        if payload is None or len(payload) < 2:
            return None
        return (payload[0] << 8) | payload[1]

    def connect_by_pin(self, pin: int) -> bool:
        """Scan for NUS devices and connect to the one matching the given PIN."""
        devices = self.scan_devices()
        for addr, name, dev_pin in devices:
            if dev_pin == pin:
                print(f"PIN {pin} matched: {name} at {addr}")
                return self.connect(address=addr)
        print(f"No device found with PIN {pin}")
        return False

    def connect(self, address: Optional[str] = None) -> bool:
        """Scan (if no address given) and connect to the device."""
        if address is None:
            address = self.scan()
        if address is None:
            return False
        try:
            return self._run_coro(
                self._connect_async(address), timeout=CONNECT_TIMEOUT_SEC + 5
            )
        except Exception as e:
            print(f"BLE connect failed: {e}")
            return False

    async def _connect_async(self, address: str) -> bool:
        # Disconnect previous client if any (e.g. reconnect after failure)
        if self._client:
            try:
                await self._client.disconnect()
            except Exception:
                pass
            self._client = None

        def on_disconnect(client):
            self._connected = False
            print("BLE disconnected.")

        self._client = BleakClient(address, disconnected_callback=on_disconnect)
        await self._client.connect(timeout=CONNECT_TIMEOUT_SEC)

        # Request larger MTU to avoid fragmentation on protocol messages
        if hasattr(self._client, "mtu_size"):
            print(f"BLE MTU: {self._client.mtu_size}")

        self._connected = True
        print(f"BLE connected to {address}")
        return True

    def send(self, data: bytes) -> bool:
        """Send data to the ESP32 via NUS RX characteristic."""
        if not self._connected or not self._client:
            return False
        try:
            self._run_coro(self._send_async(data), timeout=5.0)
            return True
        except Exception as e:
            print(f"BLE send error: {e}")
            self._connected = False
            return False

    async def _send_async(self, data: bytes):
        # Write without response for lower latency
        await self._client.write_gatt_char(NUS_RX_UUID, data, response=False)

    def disconnect(self):
        """Disconnect from the BLE device."""
        if self._client and self._connected:
            try:
                self._run_coro(self._client.disconnect(), timeout=5.0)
            except Exception:
                pass
        self._connected = False
        self._client = None

    def close(self):
        """Shut down the background event loop."""
        self.disconnect()
        if self._loop and self._loop.is_running():
            self._loop.call_soon_threadsafe(self._loop.stop)
        if self._thread:
            self._thread.join(timeout=2.0)

File: macos/PixelAgents/PixelAgents/Transport/BLETransport.swift
import Foundation
import CoreBluetooth

/// Discovered BLE device with NUS service.
struct BLEDevice: Identifiable, Hashable {
    let id: UUID              // CBPeripheral identifier
    let name: String
    let pin: UInt16?          // 4-digit PIN from manufacturer data (nil if not present)
    let peripheral: CBPeripheral

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: BLEDevice, rhs: BLEDevice) -> Bool {
        lhs.id == rhs.id
    }
}

/// CoreBluetooth NUS (Nordic UART Service) client transport.
final class BLETransport: NSObject, TransportProtocol, ObservableObject {

    // MARK: - NUS UUIDs (must match firmware ble_service.cpp)

    static let nusServiceUUID = CBUUID(string: "6E400001-B5A3-F393-E0A9-E50E24DCCA9E")
    static let nusRxUUID      = CBUUID(string: "6E400002-B5A3-F393-E0A9-E50E24DCCA9E") // write to device
    static let nusTxUUID      = CBUUID(string: "6E400003-B5A3-F393-E0A9-E50E24DCCA9E") // notify from device

    // MARK: - BLE Battery Service (BAS) UUIDs

    static let basServiceUUID = CBUUID(string: "180F")
    static let basLevelUUID   = CBUUID(string: "2A19")

    /// Manufacturer data company ID (must match firmware BLE_MFG_COMPANY_ID)
    static let mfgCompanyID: UInt16 = 0xFFFF

    /// Connect timeout — CoreBluetooth connect() has no built-in timeout.
    private static let connectTimeout: TimeInterval = 10.0

    // MARK: - Published state

    @Published private(set) var isConnected = false
    @Published private(set) var discoveredDevices: [BLEDevice] = []
    @Published private(set) var isScanning = false
    @Published private(set) var bluetoothState: CBManagerState = .unknown
    @Published private(set) var connectedDeviceName: String?
    @Published private(set) var batteryLevel: UInt8?

    /// UUID of the currently connected peripheral (for UI).
    var connectedPeripheralID: UUID? { connectedPeripheral?.identifier }

    /// UUID of the peripheral with a pending connect() (for UI).
    var pendingPeripheralID: UUID? { pendingPeripheral?.identifier }

    // MARK: - Private state

    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    private var rxCharacteristic: CBCharacteristic?

    /// Peripheral with a pending connect() call (not yet connected).
    private var pendingPeripheral: CBPeripheral?

    /// UUID of the last successfully connected peripheral (for reconnection).
    private var lastConnectedID: UUID?

    /// Timeout work item for connect attempts.
    private var connectTimeoutWork: DispatchWorkItem?

    /// Set after UUID-based reconnect times out — skip UUID and rely on scan + auto-connect.
    private var skipUUIDReconnect = false

    /// When true, next cleanupConnection preserves discoveredDevices (for manual disconnect).
    private var keepDevicesOnCleanup = false

    /// Callback when connection is lost (for auto-reconnect).
    var onDisconnect: (() -> Void)?
    var onSettingsState: ((_ payload: Data) -> Void)?

    // MARK: - Init

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    // MARK: - Scanning

    func startScanning() {
        guard centralManager.state == .poweredOn else {
            NSLog("[BLE] Cannot scan — Bluetooth state: %d", centralManager.state.rawValue)
            return
        }
        discoveredDevices.removeAll()
        centralManager.stopScan() // Ensure clean scan session (resets duplicate tracking)
        centralManager.scanForPeripherals(
            withServices: [Self.nusServiceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
        isScanning = true
        NSLog("[BLE] Scan started")
    }

    func stopScanning() {
        centralManager.stopScan()
        isScanning = false
    }

    // MARK: - Connection

    func connect(to device: BLEDevice) {
        stopScanning()
        cancelPendingConnect()
        skipUUIDReconnect = false

        NSLog("[BLE] Connecting to: %@ (id: %@)", device.name, device.id.uuidString)
        pendingPeripheral = device.peripheral
        centralManager.connect(device.peripheral, options: nil)
        startConnectTimeout()
    }

    func connectByPin(_ pin: UInt16) {
        guard !isConnected, pendingPeripheral == nil else { return }

        if let device = discoveredDevices.first(where: { $0.pin == pin }) {
            connect(to: device)
        } else if !isScanning {
            startScanning()
        }
    }

    /// Auto-reconnect: tries UUID-based reconnect first, falls back to scanning.
    /// Returns true if an active connection attempt is in progress.
    @discardableResult
    func reconnect() -> Bool {
        guard !isConnected else { return false }

        // Connection already in progress — report it
        if pendingPeripheral != nil { return true }

        // Try UUID-based reconnect (skip after first timeout — rely on scan + auto-connect)
        if let uuid = lastConnectedID, !skipUUIDReconnect {
            let peripherals = centralManager.retrievePeripherals(withIdentifiers: [uuid])
            if let peripheral = peripherals.first {
                NSLog("[BLE] Reconnecting by UUID: %@", uuid.uuidString)
                stopScanning()
                pendingPeripheral = peripheral
                peripheral.delegate = self
                centralManager.connect(peripheral, options: nil)
                startConnectTimeout()
                return true
            }
            NSLog("[BLE] Could not retrieve peripheral %@, clearing", uuid.uuidString)
            lastConnectedID = nil
        }

        // Check if a previous scan already found the device we want
        if let uuid = lastConnectedID,
           let device = discoveredDevices.first(where: { $0.id == uuid }) {
            NSLog("[BLE] Found known device in scan results, connecting")
            connect(to: device)
            return true
        }

        // Fall back to scanning — next reconnect() call will check discoveredDevices
        if !isScanning {
            NSLog("[BLE] Starting scan for reconnect")
            startScanning()
        }
        return false
    }

    func disconnect() {
        cancelPendingConnect()
        if let peripheral = connectedPeripheral {
            centralManager.cancelPeripheralConnection(peripheral)
        }
        cleanupConnection()
    }

    /// Disconnect but preserve discovered devices so the user can reconnect manually.
    func disconnectKeepDevices() {
        keepDevicesOnCleanup = true
        cancelPendingConnect()
        if let peripheral = connectedPeripheral {
            // Keep connectedPeripheral set so didDisconnectPeripheral's guard passes
            // and cleanupConnection() consumes keepDevicesOnCleanup.
            centralManager.cancelPeripheralConnection(peripheral)
        } else {
            cleanupConnection()
        }
    }

    func send(_ data: Data) -> Bool {
        guard let peripheral = connectedPeripheral,
              let rx = rxCharacteristic,
              isConnected else { return false }

        peripheral.writeValue(data, for: rx, type: .withoutResponse)
        return true
    }

    // MARK: - PIN extraction

    /// Extract 4-digit PIN from manufacturer advertising data.
    static func extractPin(from advertisementData: [String: Any]) -> UInt16? {
        guard let mfgData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data,
              mfgData.count >= 4 else { return nil }

        // Company ID is little-endian in first 2 bytes
        let companyID = UInt16(mfgData[0]) | (UInt16(mfgData[1]) << 8)
        guard companyID == mfgCompanyID else { return nil }

        // PIN is big-endian in next 2 bytes
        let pin = (UInt16(mfgData[2]) << 8) | UInt16(mfgData[3])
        guard pin >= 1000, pin <= 9999 else { return nil }
        return pin
    }

    // MARK: - Private

    private func cancelPendingConnect() {
        connectTimeoutWork?.cancel()
        connectTimeoutWork = nil
        if let pending = pendingPeripheral {
            pendingPeripheral = nil // Clear before cancel to avoid callback interference
            centralManager.cancelPeripheralConnection(pending)
        }
    }

    private func startConnectTimeout() {
        connectTimeoutWork?.cancel()
        let work = DispatchWorkItem { [weak self] in
            guard let self = self, self.pendingPeripheral != nil else { return }
            NSLog("[BLE] Connect timeout after %.0fs — falling back to scan", Self.connectTimeout)
            self.cancelPendingConnect()
            // Don't clear lastConnectedID — didDiscover needs it for auto-connect
            self.skipUUIDReconnect = true
            self.startScanning()
        }
        connectTimeoutWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + Self.connectTimeout, execute: work)
    }

    private func cleanupConnection() {
        connectedPeripheral = nil
        pendingPeripheral = nil
        rxCharacteristic = nil
        connectedDeviceName = nil
        batteryLevel = nil
        isConnected = false
        if keepDevicesOnCleanup {
            keepDevicesOnCleanup = false
        } else {
            discoveredDevices.removeAll()
        }
    }
}

// MARK: - CBCentralManagerDelegate

extension BLETransport: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        NSLog("[BLE] Central state changed: %d", central.state.rawValue)
        bluetoothState = central.state
        if central.state != .poweredOn {
            stopScanning()
            if isConnected {
                cleanupConnection()
                onDisconnect?()
            }
        }
    }

    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral,
                         advertisementData: [String: Any], rssi RSSI: NSNumber) {
        let name = peripheral.name ?? advertisementData[CBAdvertisementDataLocalNameKey] as? String ?? "Unknown"
        let pin = Self.extractPin(from: advertisementData)

        let device = BLEDevice(id: peripheral.identifier, name: name, pin: pin, peripheral: peripheral)

        if !discoveredDevices.contains(where: { $0.id == device.id }) {
            NSLog("[BLE] Discovered: %@ (PIN: %@, id: %@)", name, pin.map { String($0) } ?? "none", peripheral.identifier.uuidString)
            discoveredDevices.append(device)

            // Note: reconnect() checks discoveredDevices on each timer tick,
            // so we don't auto-connect here — that avoids state races with BridgeService.
        }
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        NSLog("[BLE] Connected: %@ (id: %@)", peripheral.name ?? "Unknown", peripheral.identifier.uuidString)
        connectTimeoutWork?.cancel()
        connectTimeoutWork = nil
        lastConnectedID = peripheral.identifier
        skipUUIDReconnect = false
        pendingPeripheral = nil
        connectedPeripheral = peripheral
        connectedDeviceName = peripheral.name ?? "PixelAgents"
        peripheral.delegate = self
        peripheral.discoverServices([Self.nusServiceUUID, Self.basServiceUUID])
    }

    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
        NSLog("[BLE] Failed to connect: %@ — %@", peripheral.name ?? "Unknown", error?.localizedDescription ?? "unknown error")
        connectTimeoutWork?.cancel()
        connectTimeoutWork = nil
        pendingPeripheral = nil
        cleanupConnection()
        onDisconnect?()
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        NSLog("[BLE] Disconnected: %@ — %@", peripheral.name ?? "Unknown", error?.localizedDescription ?? "clean disconnect")

        // Ignore disconnect callbacks for peripherals we already cancelled/abandoned.
        // cancelPendingConnect() nils pendingPeripheral before calling cancelPeripheralConnection,
        // so the resulting callback arrives when neither pending nor connected matches.
        guard peripheral === connectedPeripheral || peripheral === pendingPeripheral else {
            NSLog("[BLE] Ignoring disconnect for stale peripheral")
            return
        }

        connectTimeoutWork?.cancel()
        connectTimeoutWork = nil
        cleanupConnection()
        onDisconnect?()
    }
}

// MARK: - CBPeripheralDelegate

extension BLETransport: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard error == nil, let services = peripheral.services else {
            NSLog("[BLE] Service discovery failed: %@", error?.localizedDescription ?? "no services")
            centralManager.cancelPeripheralConnection(peripheral)
            cleanupConnection()
            onDisconnect?()
            return
        }
        for service in services {
            if service.uuid == Self.nusServiceUUID {
                peripheral.discoverCharacteristics([Self.nusRxUUID, Self.nusTxUUID], for: service)
            } else if service.uuid == Self.basServiceUUID {
                peripheral.discoverCharacteristics([Self.basLevelUUID], for: service)
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
        guard error == nil, let characteristics = service.characteristics else {
            NSLog("[BLE] Characteristic discovery failed: %@", error?.localizedDescription ?? "no characteristics")
            centralManager.cancelPeripheralConnection(peripheral)
            cleanupConnection()
            onDisconnect?()
            return
        }
        for char in characteristics {
            if char.uuid == Self.nusRxUUID {
                rxCharacteristic = char
            } else if char.uuid == Self.nusTxUUID {
                // Subscribe to notifications (future use)
                peripheral.setNotifyValue(true, for: char)
            } else if char.uuid == Self.basLevelUUID {
                // Read initial value and subscribe to battery level notifications
                peripheral.readValue(for: char)
                peripheral.setNotifyValue(true, for: char)
                NSLog("[BLE] Battery service discovered")
            }
        }

        // Mark connected once RX characteristic is found
        if rxCharacteristic != nil {
            NSLog("[BLE] NUS ready — transport connected")
            isConnected = true
        }
    }

    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        guard error == nil, let data = characteristic.value, !data.isEmpty else { return }

        if characteristic.uuid == Self.basLevelUUID {
            let level = data[0]
            if level <= 100 {
                DispatchQueue.main.async {
                    self.batteryLevel = level
                }
                NSLog("[BLE] Battery level: %d%%", level)
            }
        } else if characteristic.uuid == Self.nusTxUUID {
            // Parse protocol frame: [0xAA][0x55][type][payload...][checksum]
            guard data.count >= 9,
                  data[0] == ProtocolBuilder.syncByte1,
                  data[1] == ProtocolBuilder.syncByte2,
                  data[2] == ProtocolBuilder.msgSettingsState else { return }
            let check = data[2] ^ data[3] ^ data[4] ^ data[5] ^ data[6] ^ data[7]
            guard check == data[8] else { return }
            let payload = data.subdata(in: 3..<8)
            onSettingsState?(payload)
        }
    }
}
